File: app/src/main/java/com/qwe7002/telegram_rc/static_class/other.java
package com.qwe7002.telegram_rc.static_class;

import android.Manifest;
import android.annotation.TargetApi;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import android.telephony.SubscriptionInfo;
import android.telephony.SubscriptionManager;
import android.util.Log;

import androidx.core.app.ActivityCompat;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.qwe7002.telegram_rc.R;
import com.qwe7002.telegram_rc.data_structure.sms_request_info;

import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

import io.paperdb.Paper;


public class other {

    public static String get_nine_key_map_convert(String input) {
        final Map<Character, Integer> nine_key_map = new HashMap<Character, Integer>() {
            {
                put('A', 2);
                put('B', 2);
                put('C', 2);
                put('D', 3);
                put('E', 3);
                put('F', 3);
                put('G', 4);
                put('H', 4);
                put('I', 4);
                put('J', 5);
                put('K', 5);
                put('L', 5);
                put('M', 6);
                put('N', 6);
                put('O', 6);
                put('P', 7);
                put('Q', 7);
                put('R', 7);
                put('S', 7);
                put('T', 8);
                put('U', 8);
                put('V', 8);
                put('W', 9);
                put('X', 9);
                put('Y', 9);
                put('Z', 9);
            }
        };
        StringBuilder result_stringbuilder = new StringBuilder();
        char[] ussd_char_array = input.toUpperCase().toCharArray();
        for (char c : ussd_char_array) {
            if (Character.isUpperCase(c)) {
                result_stringbuilder.append(nine_key_map.get(c));
            } else {
                result_stringbuilder.append(c);
            }
        }
        return result_stringbuilder.toString();
    }

    public static long parse_string_to_long(@NotNull String int_str) {
        long result = 0;
        if (!int_str.isEmpty()) {
            try {
                result = Long.parseLong(int_str);
            } catch (NumberFormatException e) {
                e.printStackTrace();
            }
        }
        return result;
    }

    @NotNull
    public static String get_send_phone_number(@NotNull String phone_number) {
        phone_number = get_nine_key_map_convert(phone_number);
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < phone_number.length(); ++i) {
            char c = phone_number.charAt(i);
            if (c == '+' || Character.isDigit(c)) {
                result.append(c);
            }
        }
        return result.toString();
    }

    public static String get_dual_sim_card_display(Context context, int slot, boolean show_name) {
        String dual_sim = "";
        if (slot == -1) {
            return dual_sim;
        }
        if (other.get_active_card(context) >= 2) {
            String result = "";
            if (show_name) {
                result = "(" + get_sim_display_name(context, slot) + ")";
            }
            dual_sim = "SIM" + (slot + 1) + result + " ";
        }
        return dual_sim;
    }

    public static boolean is_phone_number(@NotNull String str) {
        for (int i = str.length(); --i >= 0; ) {
            char c = str.charAt(i);
            if (c == '+') {
                continue;
            }
            if (!Character.isDigit(c)) {
                return false;
            }
        }
        return true;
    }

    @NotNull
    @TargetApi(Build.VERSION_CODES.N)
    public static String get_data_sim_id(Context context) {
        String result = "Unknown";
        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_PHONE_STATE) != PackageManager.PERMISSION_GRANTED) {
            Log.d("get_data_sim_id", "No permission.");
            return result;
        }
        SubscriptionManager subscriptionManager = (SubscriptionManager) context.getSystemService(Context.TELEPHONY_SUBSCRIPTION_SERVICE);
        assert subscriptionManager != null;
        SubscriptionInfo info = subscriptionManager.getActiveSubscriptionInfo(SubscriptionManager.getDefaultDataSubscriptionId());
        if (info == null) {
            return result;
        }
        result = String.valueOf(info.getSimSlotIndex() + 1);
        return result;
    }


    @NotNull
    public static Long get_message_id(String result) {
        JsonObject result_obj = JsonParser.parseString(result).getAsJsonObject().get("result").getAsJsonObject();
        return result_obj.get("message_id").getAsLong();
    }

    @NotNull
    public static Notification get_notification_obj(@NotNull Context context, String notification_name) {
        NotificationChannel channel = new NotificationChannel(notification_name, notification_name,
                NotificationManager.IMPORTANCE_MIN);
        NotificationManager manager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        assert manager != null;
        manager.createNotificationChannel(channel);
        Notification.Builder notification = new Notification.Builder(context, notification_name).setAutoCancel(false)
                .setSmallIcon(R.drawable.ic_stat)
                .setOngoing(true)
                .setTicker(context.getString(R.string.app_name))
                .setContentTitle(context.getString(R.string.app_name))
                .setContentText(notification_name + context.getString(R.string.service_is_running));
        return notification.build();
    }

    public static int get_sub_id(Context context, int slot) {
        int active_card = other.get_active_card(context);
        if (active_card >= 2) {
            if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_PHONE_STATE) != PackageManager.PERMISSION_GRANTED) {
                Log.i("get_sub_id", "get_sub_id: No permission");
                return -1;
            }
            SubscriptionManager subscriptionManager = (SubscriptionManager) context.getSystemService(Context.TELEPHONY_SUBSCRIPTION_SERVICE);
            assert subscriptionManager != null;
            return subscriptionManager.getActiveSubscriptionInfoForSimSlotIndex(slot).getSubscriptionId();
        }
        return -1;
    }

    public static int get_active_card(Context context) {
        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_PHONE_STATE) != PackageManager.PERMISSION_GRANTED) {
            return -1;
        }
        SubscriptionManager subscriptionManager = (SubscriptionManager) context.getSystemService(Context.TELEPHONY_SUBSCRIPTION_SERVICE);
        assert subscriptionManager != null;
        return subscriptionManager.getActiveSubscriptionInfoCount();
    }

    public static String get_sim_display_name(Context context, int slot) {
        final String TAG = "get_sim_display_name";
        String result = "Unknown";
        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_PHONE_STATE) != PackageManager.PERMISSION_GRANTED) {
            Log.i(TAG, "No permission.");
            return result;
        }
        SubscriptionManager subscriptionManager = (SubscriptionManager) context.getSystemService(Context.TELEPHONY_SUBSCRIPTION_SERVICE);
        assert subscriptionManager != null;
        SubscriptionInfo info = subscriptionManager.getActiveSubscriptionInfoForSimSlotIndex(slot);
        if (info == null) {
            Log.d(TAG, "The active card is in the second card slot.");
            if (get_active_card(context) == 1 && slot == 0) {
                info = subscriptionManager.getActiveSubscriptionInfoForSimSlotIndex(1);
            }
            if (info == null) {
                Log.d(TAG, "get_sim_display_name: Unable to obtain information");
                return result;
            }
            return result;
        }
        result = info.getDisplayName().toString();
        if (info.getDisplayName().toString().contains("CARD") || info.getDisplayName().toString().contains("SUB")) {
            result = info.getCarrierName().toString();
        }
        return result;
    }


    public static void add_message_list(long message_id, String phone, int slot) {
        sms_request_info item = new sms_request_info();
        item.phone = phone;
        item.card = slot;
        Paper.book().write(String.valueOf(message_id), item);
        Log.d("add_message_list", "add_message_list: " + message_id);
    }

}

File: app/src/main/java/com/qwe7002/telegram_rc/static_class/Other.kt
package com.qwe7002.telegram_rc.static_class

import android.Manifest
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.telephony.SubscriptionManager
import android.util.Log
import androidx.core.app.ActivityCompat
import com.google.gson.Gson
import com.google.gson.JsonParser
import com.qwe7002.telegram_rc.MMKV.Const
import com.qwe7002.telegram_rc.R
import com.qwe7002.telegram_rc.data_structure.SMSRequestInfo
import com.tencent.mmkv.MMKV
import java.util.Locale

object Other {
    private val NINE_KEY_MAP = mapOf(
        'A' to '2', 'B' to '2', 'C' to '2',
        'D' to '3', 'E' to '3', 'F' to '3',
        'G' to '4', 'H' to '4', 'I' to '4',
        'J' to '5', 'K' to '5', 'L' to '5',
        'M' to '6', 'N' to '6', 'O' to '6',
        'P' to '7', 'Q' to '7', 'R' to '7', 'S' to '7',
        'T' to '8', 'U' to '8', 'V' to '8',
        'W' to '9', 'X' to '9', 'Y' to '9', 'Z' to '9'
    )

    fun getNineKeyMapConvert(input: String): String {
        return input.uppercase(Locale.ROOT).map { c ->
            NINE_KEY_MAP[c] ?: c
        }.joinToString("")
    }

    @JvmStatic
    fun parseStringToLong(intStr: String): Long {
        var result: Long = 0
        if (intStr.isNotEmpty()) {
            try {
                result = intStr.toLong()
            } catch (e: NumberFormatException) {
                e.printStackTrace()
            }
        }
        return result
    }

    @JvmStatic
    fun getSendPhoneNumber(phoneNumber: String): String {
        var convertNumber = phoneNumber
        convertNumber = getNineKeyMapConvert(convertNumber)
        val result = StringBuilder()
        for (element in convertNumber) {
            if (element == '+' || Character.isDigit(element)) {
                result.append(element)
            }
        }
        return result.toString()
    }

    @JvmStatic
    fun getDualSimCardDisplay(context: Context, slot: Int): String {
        var dualSim = ""
        if (slot == -1) {
            return dualSim
        }
        if (getActiveCard(context) >= 2) {
            dualSim = "SIM" + (slot + 1) + " "
        }
        return dualSim
    }

    @JvmStatic
    fun isPhoneNumber(str: String): Boolean {
        var i = str.length
        while (--i >= 0) {
            val c = str[i]
            if (c == '+') {
                continue
            }
            if (!Character.isDigit(c)) {
                return false
            }
        }
        return true
    }

    @JvmStatic
    fun getDataSimId(context: Context): String {
        var result = "Unknown"
        if (ActivityCompat.checkSelfPermission(
                context,
                Manifest.permission.READ_PHONE_STATE
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            Log.d("get_data_sim_id", "No permission.")
            return result
        }
        val subscriptionManager =
            (context.getSystemService(Context.TELEPHONY_SUBSCRIPTION_SERVICE) as SubscriptionManager)
        val info =
            subscriptionManager.getActiveSubscriptionInfo(SubscriptionManager.getDefaultDataSubscriptionId())
                ?: return result
        result = (info.simSlotIndex + 1).toString()
        return result
    }


    @JvmStatic
    fun getMessageId(result: String?): Long {
        Log.d("getMessageId", "getMessageId: "+ result)
        val resultObj = JsonParser.parseString(result).asJsonObject["result"].asJsonObject
        return resultObj["message_id"].asLong
    }

    @JvmStatic
    fun getNotificationObj(context: Context, notificationName: String): Notification.Builder {
        val manager =
            (context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager)
        val channelExists = manager.getNotificationChannel(notificationName) != null
        if (!channelExists) {
            val channel = NotificationChannel(
                notificationName, notificationName,
                NotificationManager.IMPORTANCE_MIN
            )
            manager.createNotificationChannel(channel)
        }
        val appName = context.getString(R.string.app_name)
        val builder = Notification.Builder(context, notificationName).setAutoCancel(false)
            .setSmallIcon(R.drawable.ic_stat)
            .setOngoing(true)
            .setTicker(appName)
            .setContentTitle(appName)
            .setContentText(notificationName + context.getString(R.string.service_is_running))
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            builder.setForegroundServiceBehavior(Notification.FOREGROUND_SERVICE_IMMEDIATE)
        }
        return builder
    }

    @JvmStatic
    fun getSubId(context: Context, slot: Int): Int {
        val activeCard = getActiveCard(context)
        if (activeCard >= 2) {
            if (ActivityCompat.checkSelfPermission(
                    context,
                    Manifest.permission.READ_PHONE_STATE
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                Log.i("getSubId", "get_sub_id: No permission")
                return -1
            }
            val subscriptionManager =
                (context.getSystemService(Context.TELEPHONY_SUBSCRIPTION_SERVICE) as SubscriptionManager)
            return subscriptionManager.getActiveSubscriptionInfoForSimSlotIndex(slot).subscriptionId
        }
        return -1
    }

    @JvmStatic
    fun getActiveCard(context: Context): Int {
        if (ActivityCompat.checkSelfPermission(
                context,
                Manifest.permission.READ_PHONE_STATE
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            return -1
        }
        val subscriptionManager =
            (context.getSystemService(Context.TELEPHONY_SUBSCRIPTION_SERVICE) as SubscriptionManager)
        return subscriptionManager.activeSubscriptionInfoCount
    }

    @JvmStatic
    fun getSimDisplayName(context: Context, slot: Int): String {
        val TAG = "get_sim_display_name"
        var result = "Unknown"
        if (ActivityCompat.checkSelfPermission(
                context,
                Manifest.permission.READ_PHONE_STATE
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            Log.i(TAG, "No permission.")
            return result
        }
        val subscriptionManager =
            (context.getSystemService(Context.TELEPHONY_SUBSCRIPTION_SERVICE) as SubscriptionManager)
        var info = subscriptionManager.getActiveSubscriptionInfoForSimSlotIndex(slot)
        if (info == null) {
            Log.d(TAG, "The active card is in the second card slot.")
            if (getActiveCard(context) == 1 && slot == 0) {
                info = subscriptionManager.getActiveSubscriptionInfoForSimSlotIndex(1)
            }
            if (info == null) {
                Log.d(TAG, "get_sim_display_name: Unable to obtain information")
                return result
            }
        }

        val displayName = info.displayName?.toString() ?: ""
        result = if (displayName.contains("CARD") || displayName.contains("SUB")) {
            info.carrierName?.toString() ?: displayName
        } else {
            displayName
        }
        return result
    }


    fun addMessageList(messageId: Long, phone: String, slot: Int) {
        val item = SMSRequestInfo()
        item.phone = phone
        item.card = slot
        MMKV.mmkvWithID(Const.CHAT_INFO_MMKV_ID)
            .putString(messageId.toString(), Gson().toJson(item))
        Log.d("add_message_list", "add_message_list: $messageId")
    }
}
